### PR TITLE
Feature/dependency repo caching

### LINKED ISSUE #694: [FEATURE] Cache repositories
### Feature summary

Allow viash to cache repositories within the same viash ns build, but potentially also outside of it

### Feature description

Right now, using the same repository in multiple dependencies seems to cause Viash to clone the repository multiple times. Ideally, the repository should only need to be cloned once per viash ns build, or even fewer times if we allow cross-build caching.

### Why is this feature beneficial?

Large repositories might take a while to clone.

### Alternatives considered

_No response_

### Possible solution

_No response_

### Confirmation

- [X] I have searched the existing issues to make sure this is not a duplicate request.
- [X] I have provided clear and concise information about the feature.

### PR DESCRIPTION
**Pull request by Grifs**
## Related issue(s)

Closes #694

## Type of Change

- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] New functionality (non-breaking change which adds functionality)
- [ ] Major change (non-breaking change which modifies existing functionality)
- [ ] Minor change (non-breaking change which does not modify existing functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] This change requires a documentation update

## Checklist

Requirements:

- [x] I have read the [CONTRIBUTING](CONTRIBUTING.md) doc.
- [x] I have performed a self-review of my code by checking the "Changed Files" tab.
- [x] My code follows the code style of this project.

Tests:

- [ ] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.

Documentation:

- [x] Proposed changes are described in the CHANGELOG.md.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: src/main/scala/io/viash/config/dependencies/AbstractGitRepository.scala [hunk 1]
```diff
@@ -59,22 +66,75 @@ trait AbstractGitRepository extends Repository with Logging {
...
+      case 0 => res.output.split("\t").head
+      case _ => "Remote hash not found"
+    }
+  }
+
   def getCheckoutUri(): String
+  def getCacheIdentifier(): Option[String]
+  def fullCachePath: Option[Path] = {
+    val cacheIdentifier = getCacheIdentifier()
+    cacheIdentifier.map(cacheIdentifier => Paths.get(SysEnv.viashHome).resolve("cache").resolve(cacheIdentifier))
```
Comment (rcannood) [inline src/main/scala/io/viash/config/dependencies/AbstractGitRepository.scala:112]: I'd be inclined to propose to cache them to `~/.viash/repositories/git-...`, or `repository_cache`.

```suggestion
    cacheIdentifier.map(cacheIdentifier => Paths.get(SysEnv.viashHome).resolve("repositories").resolve(cacheIdentifier))
```

**Review by rcannood — COMMENTED** [review]
This looks like great stuff!

Some questions:

* How much of the logging is sent to STDOUT/STDERR by default?
* Is there anything we can already test?

File: src/main/scala/io/viash/config/dependencies/AbstractGitRepository.scala (hunk 1 above)
Comment (Grifs) [reply src/main/scala/io/viash/config/dependencies/AbstractGitRepository.scala:112]: tomato potato ;) ... but sure!

**Comment by rcannood** [comment]
> I changed it to 1 line, e.g.
>
> Fetching repo for https://viash-hub.com/hendrik/dependency_test2.git
>
> as I re-discovered the --loglevel xxx feature 🤭

that's nice!

> I moved the low level git calls to the helper.Git object, refactored a bit and added tests for those.
>
> I'll see what I can do about testing the caching as a whole. I might have an idea or two.

:+1!

**Review by rcannood — APPROVED** [review]

**Pull request merged**